### ISSUE TITLE
More flexible expression for PARTITION BY

### ISSUE THREAD

**Issue by olegkv** [ClickHouse/ClickHouse#12436]
**Use case**

- Data Warehouse with many tables that have data coming from various sources.

- All tables are time-based, they all have `DateTime` or `Date` column.

- Data comes in batches, aligned by various date parts, it can be a batch for an hour or batch for a day, 
it depends on type of information stored in each particular table.

- Each table has a particular fixed batch scheme, one table has hourly batches, another table has daily batches. Each table can have only one batch type.

- Each batch for a particular table is expected to completely override the previous batch for the same time range. 
Rows in a batch are not necessarily the same, new batch for the same time period can have a different set of rows.

- Batches for the same time range can re-imported many times, it is like real-time adjustments for a particular time period.

- For some "hot" time (usually a few days, but can be up to 2 weeks), a given time period is considered "hot", 
and batches with adjustments (to override previously imported batches) come very frequently.

- After "hot" time elapsed, given time range becomes "cold" and it is not updated anymore.

As an example, consider some trading system, where
- some preliminary sales records are exported to Data Warehouse in hourly batches, one batch for each physical hour,
- initial batch comes immediately after a physical hour is finished, and then adjusted batches for the same hour
come multiple times during a period of few days after the initial batch.  It is like some rows in a batch need time for approval or adjustment by some higher authority.
- Data Warehouse users want to run queries to see original batches immediately, and they expect that with time, 
rows will be fully corrected (approved, adjusted) and after that, they'll see final results.
- But before final results settle, users want to see preliminary results immediately as they are available.
- After all information for given physical hour is settled, usually in few days, batches for that physical hour stop coming.

So, ETL logic implies that data for a batch time period needs to be deleted before that batch can be re-imported.

Using `ReplacingMergeTree` for such a scenario does not seem convenient, the ETL process does not track row versions and it would be too expensive to make it do so.

Using ALTER TABLE DELETE before importing each batch also does not seem convenient, because mutations would be too frequent for "hot" periods and thus too expensive.

What seems more convenient, is to partition data by the hour or day, 
depending on batch size requirement for a particular table and just delete a partition before re-importing a batch.
But this would create too many parts, monthly partitions are much more optimal for storing long history.

**Proposed solution**

What seems convenient is partition expressions supporting variable partition ranges, 
for instance, one expression for "cold" data, one expression for "hot" data and the additional expression
to differentiate hot time range from cold time range.

Something like this for hourly batches:
```
PARTITION BY toYYYYMM(TransactionTime)
   HOT_RANGE BY toHour(TransactionTime)
   HOT_RANGE_LIMIT(TransactionTime) 48 HOURS
```
or this for daily batches:
```
PARTITION BY toYYYYMM(TransactionTime)
    HOT_RANGE BY YYYYMMDD(TransactionTime)
    HOT_RANGE_LIMIT(TransactionTime) 14 DAYS
```
Or, alternatively, partition expression like:
PARTITION BY Expr1 LIMIT N1, Expr2 LIMIT N2, Expr3 LIMIT N3 ...
where limits are for max number of partitions for each expression.

With such a feature, re-importing hourly or daily batch would use simple and fast DROP PARTITION calls and then re-insert new data as usual.

MergeTree engine would keep hourly (or daily or other time range) partitions for a predefined time (or up to provided max number of partitions for each expression) and then merge them into larger "cold" partitions.

Also, such a feature could (hopefully) allow optimizing part merges for "hot" periods so that large parts are not created for data which is still hot and can still be changed.

**Alternatives considered**

Using ReplacingMergeTree, using table mutations

**Comment by alex-zaitsev** [comment]
Related #8089

**Comment by hodgesrm** [comment]
This feature has been requested by a number of users. Would it work to tie the partition scheme to the underlying storage type?  I'm a little concerned that adding unconnected partitioning rules as suggested above adds a lot of complexity to storage implementation, which in turn puts the burden on DBAs to make good decisions.  We already have rules implemented for storage tiers.

**Comment by olegkv** [comment]
Yes, I think it could work. 
If I understand correctly, the idea is that partition expression would be specified per volume in storage policy? If so, one volume would contain daily partitions and other volume would contain monthly partitions. 
'move_factor' in storage policy could be specified not only as a percentage of volume size, but also as a maximum number of partitions allowed per volume, so new-coming partitions would push older partitions to other volume.
The important point is that, after a partition is moved to a volume with different partition expression, it will merge with other partitions to match that expression. So, if daily partition is moved to a volume where only monthly partitions exist, it will disappear and its data will be merged with other data for same month.
All that seems good and convenient to me, unless I miss something..

**Comment by hodgesrm** [comment]
That's the general idea.  

The other part of this is that as you move between volumes you want to control number of parts as well as background merges, on them, which is very closely related to what you are doing.  Slower storage should merge less of not at all, as this is expensive and performs poorly, whereas reads can be speeded up through caching.  Moreover, the fewer parts that ClickHouse needs to scan, the faster your queries will go.

**Comment by olegkv** [comment]
As far as I understand, using different partition expressions for different volumes kinda solves that problem too (number and frequency of background merges, write amplification, etc), because new data comes to hot storage only where partitions are granular enough, and cold storage receives data only as already ordered partitions, pushed out from hot storage by new data. So, if such feature is enabled, it seems like it is going to indirectly solve write amplification problems too..

**Comment by olegkv** [comment]
@hodgesrm something for your consideration, unrelated to storage policies, looks like supporting general partition expressions like:
PARTITION BY Expr1 LIMIT N1, Expr2 LIMIT N2, Expr3 LIMIT N3 ...
where new data always comes to partitions controlled by first expression and then merged into next level only when partition number limit is reached, can solve many write amplification problems.

For example, it can be:
PARTITION BY toYYYYMMDD(date) LIMIT 10, toYYYYMM(date) LIMIT 6, toYYYY(date)
that means hottest data will be stored in small daily partitions, when limit (10 daily partitions) is reached, data is allowed to be merged into less hot monthly partitions and when monthly limit (6 months) is reached, data is allowed to be merged into cold large annual partitions. In such case write amplification can be reduced to minimum, as far as I see, because hottest data is stored in small parts and is moved to larger parts only after it "cools down" enough.
Just something for you to consider, and adding such feature tied to storage policies is good for me too.

**Comment by alexey-milovidov** [comment]
This is difficult to implement.
